src/modules/flow/iio/nodes.c

```diff
@@ -25,43 +25,117 @@
 
 #include <sol-iio.h>
 
-struct gyroscope_data {
+enum iiodevice_types {
+    IIO_GYROSCOPE=1,
+    IIO_MAGNETOMETER=2,
+    IIO_THERMOMETER=3,
+    IIO_PRESSURE_SENSOR=4,
+    IIO_COLOR=5,
+    IIO_ACCELEROMETER=6,
+    IIO_HUMIDITY_SENSOR=7,
+    IIO_ADC=8,
+    IIO_LIGHT_SENSOR=9,
+    IIO_PROXIMITY=10,
+};
+
+struct iiodevice_out {
+    union {
+        struct sol_direction_vector direction;
+        struct sol_drange drange;
+        struct sol_rgb rgb;
+    };
+};
+
+#define MAX_CHANNEL			(8)
+
+struct iiodevice_data {
     struct sol_iio_config config;
-    struct sol_direction_vector scale;
-    struct sol_direction_vector offset;
+    union {
+        struct sol_direction_vector scale_vector;
+        double scales[MAX_CHANNEL];
+    };
+
+    union {
+        struct sol_direction_vector offset_vector;
+        double offsets[MAX_CHANNEL];
+    };
+ 
     struct sol_drange_spec out_range;
     struct sol_iio_device *device;
-    struct sol_iio_channel *channel_x;
-    struct sol_iio_channel *channel_y;
-    struct sol_iio_channel *channel_z;
+    struct sol_iio_channel *channels[MAX_CHANNEL];
+    enum iiodevice_types type;
+    uint8_t channel_count;
     bool buffer_enabled : 1;
     bool use_device_default_scale : 1;
     bool use_device_default_offset : 1;
 };
 
+struct iio_node_type {
+    struct sol_flow_node_type base;
+    uint16_t out_port;
+};
+
 static void
 reader_cb(void *data, struct sol_iio_device *device)
 {
     static const char *errmsg = "Could not read channel buffer values";
     struct sol_flow_node *node = data;
-    struct gyroscope_data *mdata = sol_flow_node_get_private_data(node);
-    struct sol_direction_vector out = {
-        .min = mdata->out_range.min,
-        .max = mdata->out_range.max
-    };
+    struct iiodevice_data *mdata = sol_flow_node_get_private_data(node);
+    struct iiodevice_out out;
+    struct iio_node_type *type;
+    double tmp;
     bool b;
 
-    b = sol_iio_read_channel_value(mdata->channel_x, &out.x);
-    if (!b) goto error;
-
-    b = sol_iio_read_channel_value(mdata->channel_y, &out.y);
-    if (!b) goto error;
-
-    b = sol_iio_read_channel_value(mdata->channel_z, &out.z);
-    if (!b) goto error;
+    type = (struct iio_node_type *)sol_flow_node_get_type(node);
+
+    switch (mdata->type) {
```

Comment:
why don't you break this into multiple functions, one for each kind of type (ie: read_direction_vector to be used by gyroscope, magnetometer and accelerometer). Then you set this to `type` via `extra_methods` on your json.